io-sim/test/Test/IOSim.hs

```diff
@@ -773,3 +793,203 @@ selectTraceSay :: Trace a -> [String]
+      go s =
+        if s <= 1
+          then Leaf <$> gen
+          else Branch <$> go (pred s) <*> go (pred s)
+
+instance Arbitrary a => Arbitrary (BinaryTree a) where
+    arbitrary = sized $ \s -> genBinaryTree s arbitrary
+
+    shrink Leaf{} = []
+    shrink (Branch left right) =
+      [ Branch left' right
+      | left' <- shrink left
+      ] ++
+      [ Branch left right'
+      | right' <- shrink right
+      ]
+
+-- A skewed arbitrary generator of boolean values
+newtype SkewedBool = SkewedBool { runSkewedBool :: Bool }
+    deriving Show
+
+instance Arbitrary SkewedBool where
+    arbitrary = frequency 
+      [ (1, pure (SkewedBool True))
+      , (2, pure (SkewedBool False))
+      ]
+    shrink (SkewedBool b) = map SkewedBool (shrink b)
+
+-- |
+-- A tree of ids and boolean values, where ids are positive smaller or equal 5.
+-- The list contains initial values.
+--
+data LabeledTree = LabeledTree (BinaryTree (Int, Bool)) [Int]
+    deriving Show
+
+-- |
+-- Generate a small integer number from a constant distribution.
+--
+genInt :: Int -> Gen Int
+genInt l = oneof $ map return [0 .. pred l]
+
+instance Arbitrary LabeledTree where
+    arbitrary = do
+        -- size (max number of variables) between 1 and 5
+        s <- succ <$> genInt 5
+        tree <- genBinaryTree 10 $
+              (,)
+          <$> genInt s
+          <*> (runSkewedBool <$> arbitrary)
+        x0 <- replicateM s arbitrary
+        return $ LabeledTree tree x0
+
+    shrink (LabeledTree tree xs) =
+      [ LabeledTree tree' xs
+      | tree' <- shrink tree
+      ]
+
+valid_Labeled_Tree :: LabeledTree -> Bool
+valid_Labeled_Tree (LabeledTree tree xs) = all (\(x, _) -> x < length xs) tree
+
+prop_arbitrary_LabeledTree :: LabeledTree -> Bool
+prop_arbitrary_LabeledTree t@(LabeledTree tree xs) =
+     valid_Labeled_Tree t
+  && length (nub $ foldr (\(y,_) ys -> y:ys) [] tree) <= length xs -- or 5
+
+prop_shrink_LabeledTree :: LabeledTree -> Bool
+prop_shrink_LabeledTree l = all valid_Labeled_Tree (shrink l)
+
+prop_nested_orElse :: Fun Int Int
+                   -> LabeledTree
+                   -> Property
+prop_nested_orElse f (LabeledTree tree xs) =
+    tabulate "depth" [show treeDepth] $
+    tabulate "number of variables" [show $ length $ nub $ foldr (\(y,_) ys -> y:ys) [] tree] $
+    case runSim (exp_ :: forall s. SimM s [Int]) of
+      Left e@FailureDeadlock ->
+        -- termination condition
+        if any snd tree
+          then counterexample (show e) False
+          else property True
+      Left  e -> counterexample (show e) False
+      Right r -> ioProperty $ (r ===) <$> (exp_ :: IO [Int])
+  where
+    g :: MonadSTM m => [TVar m Int] -> (Int, Bool) -> STM m ()
+    g vs (x, True)  = modifyTVar (vs !! x) (applyFun f)
+    g vs (x, False) = modifyTVar (vs !! x) (applyFun f) >> retry
+
+    exp_ :: MonadSTM m => m [Int]
+    exp_ = atomically $ do
+      vs <- traverse newTVar xs
+      foldl1 orElse (g vs <$> tree)
```

Comment:
Wait, this isn't applying `orElse` according to the binary tree structure, it's applying it left nested to the flattened tree. This needs to be a structural fold, with `orElse` in the branches and `g` in the leaves.

```
foldTree leaf _branch (Leaf a) = l a
foldTree leaf branch (Branch l r) = branch (foldTree leaf branch l) (foldTree leaf branch r)
```
```
    exp_ :: MonadSTM m => m [Int]
    exp_ = atomically $ do
      vs <- traverse newTVar xs
      foldTree orElse (g vs) tree
      traverse readTVar vs 
```

Reply:
Good catch, at some point I simplified the tests too much.